### ISSUE TITLE
sorting by type - is it possible at all?

### ISSUE THREAD

**Issue by skrobul** [de-vri-es/weechat-autosort#4]
Hi!

I'm struggling with sorting rules configuration and was wondering if anyone can confirm if this is possible to setup with this plugin at all.

I'm trying to achieve following buffers order:
1. All private chats/queries first
2. Next, all channels starting with `#` (regardless of the network, or ordered by network - don't care)
3. Next, all channels starting with `&`
4. Next, all `irc.server` buffers
5. Lastly the `core` window

My setup is as follows:
1. Three IRC networks/servers, one of them is bitlbee
2. Following rules configured:

```
0: irc.*.#* = 3
1: irc.*.&* = 4
2: irc.*.* = 2
3: irc.server = 15
4: core = 9999999999
```

`group_irc` is set to `off`.

Currently, above rules result in following order: [server buffers, 1st network, 2nd network, core window] which is not what I want.

Guessing from the source code, this does not work because python stops comparison on the first or second component, regardless how the last component is scored. For example, the `sort()` method receives following data:

```
12:25 autosort: Unsorted:
[[(9999999999, u'core'), (999999999, u'weechat')],
[(999999999, u'irc'), (15, u'server'), (999999999, u'bitlbee')],
[(999999999, u'irc'), (15, u'server'), (999999999, u'net1')],
 [(999999999, u'irc'), (15, u'server'), (999999999, u'znc-freenode')],
 [(999999999, u'irc'), (999999999, u'bitlbee'), (3, u'#twitter_abc')],
 [(999999999, u'irc'), (999999999, u'bitlbee'), (3, u'&bitlbee')],
 [(999999999, u'irc'), (999999999, u'bitlbee'), (3, u'&jabber')],
 [(999999999, u'irc'), (999999999, u'net1'), (3, u'#acsas')],
 [(999999999, u'irc'), (999999999, u'net1'), (3, u'#netdevops')],
 [(999999999, u'irc'), (999999999, u'net1'), (999999999, u'priv1')],
 [(999999999, u'irc'), (999999999, u'znc-freenode'), (3, u'#linux')],
 [(999999999, u'irc'), (999999999, u'znc-freenode'), (3, u'#salt')],
 [(999999999, u'irc'), (999999999, u'znc-freenode'), (3, u'#vim')],
 [(999999999, u'irc'), (999999999, u'znc-freenode'), (999999999, u'^priv3^')],
 [(999999999, u'irc'), (999999999, u'znc-freenode'), (999999999, u'someoneelse')]]
```

For reference, this was produced by adding following debug code:

``` python
def on_buffers_changed(*args, **kwargs):
    ''' Called whenever the buffer list changes. '''
    buffers = get_buffers()
    skey = buffer_sort_key(config.rules)
    unsorted = [skey(buf) for buf in buffers]
    log("Unsorted: {0}".format(unsorted))
    buffers.sort(key=buffer_sort_key(config.rules))
    for buf in buffers:
        log("result for %s: %s" % (buf, skey(buf)))
    apply_buffer_order(buffers)
    return weechat.WEECHAT_RC_OK
```

Is there any way to avoid it?

**Comment by de-vri-es** [comment]
You are right, currently sorting is always topological based on the buffer names split on dots, meaning that the network is always considered before the name of a channel.

I'm interested in allowing more general sorting methods, but I'm not sure how to expose that without requiring a lot of knowledge on part of the user.

I was thinking to allow a custom format string list of fields to sort on, rather than always using the full buffer name split on periods.

For example, in your scenario you might configure a format/field list "{type}.{full_buffer_name}", although I'm not sure if there is something in the buffer information that can be used as a {type} field.

If you have a suggestion and/or PR I'm all ears. I'll also think about this some more.

Another option that could provide you with the behaviour you want is to allow disabling (or customizing) splitting of the buffer name. You can't avoid sorting by network that way (unless replacement patterns are extended with glob/regex support), but it would allow you to flatten the tree into a list, meaning your sorting rules would give the wanted result (still sorted per network within each matching rule).

**Comment by skrobul** [comment]
Thanks for confirmation, thought I was going crazy :)

It's probably too much of the behavior change for your plugin, but I was wondering why the splitting by dot is needed at all? What if you allowed the rules to just contain simple regular expressions and allow scoring based on that?

For example to solve this particular case, rules would be:

```
0: /^irc\.(net1|bitlbee|znc)\.[a-zA-Z]/  = 0      # All private chats/queries first
1: /^irc\.(net1|bitlbee|znc)\.#/  = 1             # Next, all channels starting with # 
2: /^irc\.(net1|bitlbee|znc)\.&/  = 2             # Next, all channels starting with &
3: /^irc\.server/ = 3                             # Next, all irc.server buffers
4: /^core/ = 4                                    # Lastly the core window
```

Similarly, for the [project's example](https://github.com/de-vri-es/weechat-autosort/blob/master/README.md#example), the rules would be:

```
0: /^core/ = 0                          # core buffer first
1: /^irc\.server\..*\.#/ = 1            # channel buffers
2: /^irc\.server\..*\.*/ = 0            # queries, etc
3: /^irc/  = 2                          # irc buffers
4: /.*/    = 1                          # everything else

```

**Comment by de-vri-es** [comment]
Without splitting on dot you would have to repeat rules for each network to achieve a sorted tree (which is the default, assuming server buffers are not merged). Otherwise all channel buffers would be grouped together instead of grouped below their respective server buffer.

Still, I do want to support more use cases such as yours. In the mean time you can of course adapt the script to your personal needs by editing the source, if you wish.

**Comment by skrobul** [comment]
Makes perfect sense - I completely forgot about the fact people use it this way - my buffer list is in the horizontal bottom window without any indentation. 

I'll try to modify the plugin when I have some spare time, but before that I'm going to try experimenting with replacements - maybe I can just replace `irc.net1`, `irc.bitlbee` and so on with empty strings thus removing any information about the networks.

**Comment by de-vri-es** [comment]
Luckily replacements are done before splitting, so you might get quite far with those.

I still like the idea of a custom sort format, but I'll have to think about that a bit more.

**Comment by mbuelte** [comment]
@skrobul Did you find a working solution to your problem using the replacements? I tried it also and I didn't got it work.

**Comment by skrobul** [comment]
@mbuelte sorry, tbh I don't remember what was the outcome...Don't think I found the solution unfortunately

**Comment by mbuelte** [comment]
Damn, thank you for your response. I need a handy solution to move merged pv_buffers on position 2. Maybe @de-vri-es  had enough time to think about implement a custom sort method based on regex.

**Comment by de-vri-es** [comment]
The current version of autosort doesn't support it. I'm working on a different system in the v3 branch [1] that uses replacement patterns (like the new built-in buffers plugin).

That might be able to do what you want. You'd still have to configure the sort rules and probably find out which signal should trigger a resort. Currently it doesn't migrate well from the old autosort. It's still a work in progress.

[1] https://github.com/de-vri-es/weechat-autosort/tree/v3

**Comment by mbuelte** [comment]
Great! Thank your for your work, I'll definitely have a look at it.

**Comment by de-vri-es** [comment]
Thanks for testing, very good to hear :D

I merged v3 back into the master branch (though it's still beta for now). It uses different config keys now from the old version to prevent breaking horribly on existing v2.x configurations. That means that if you upgrade now you will probably loose your custom rules (but luckily you got them posted here ^^).

I also tweaked the default helpers and rules a bit to showcase the feature a bit more (the rule list is practically human readable now, by default).

Additionally, I added a `${info:autosort_replace,from,to,text}` to support replacing substrings (like double hash signs: `${info:autosort_replace,##,#,${buffer.name}}`.

Bad upgrade behaviour and lacking substring replacement were pretty much the two problems blocking a release. With that fixed I intend to send it to the official script repository soon :)
